pil-analyzer/src/type_inference.rs

```diff
@@ -0,0 +1,675 @@
+                assert!(scheme.vars.is_empty());
+                self.substitute(&mut scheme.ty);
+                assert!(scheme.ty.is_concrete_type());
+                (name, scheme.ty)
+            })
+            .collect())
+    }
+
+    /// Returns, for each name declared with a type scheme, a mapping from
+    /// the type variables used by the type checker to those used in the declaration.
+    fn infer_types_inner<T: FieldElement>(
+        &mut self,
+        definitions: &mut HashMap<String, (Option<TypeScheme>, Option<&mut Expression<T>>)>,
+        expressions: &mut [(&mut Expression<T>, ExpectedType)],
+    ) -> Result<HashMap<String, HashMap<String, Type>>, String> {
+        // TODO in order to fix type inference on recursive functions, we need to:
+        // - collect all groups of functions that call each other recursively
+        // - analyze each such group in an environment, where their type schemes
+        //   are instantiated once at the start and not anymore for the symbol lookup.
+
+        // Sort the names such that called names occur first.
+        let names = sort_called_first(
+            definitions
+                .iter()
+                .map(|(n, (_, v))| (n.as_str(), v.as_deref())),
+        );
+
+        // Remove builtins from definitions and check they types are correct.
+        for (name, ty) in builtin_schemes() {
+            if let Some((_, (Some(defined_ty), _))) = definitions.remove_entry(name) {
+                assert!(
+                    ty == &defined_ty,
+                    "Invalid type for built-in scheme {name}: {}",
+                    format_type_scheme_around_name(name, &Some(defined_ty))
+                );
+            }
+        }
+
+        self.declared_types = builtin_schemes().clone();
+        // Add types from declarations. Type schemes are added without instantiating.
+        for (name, def) in definitions.iter() {
+            // This stores an (uninstantiated) type scheme for symbols with a declared
+            // polymorphic type and it creates a new (unquantified) type variable for
+            // symbols without declared type. This forces a single concrete type for the latter.
+            let ty = def.0.clone().unwrap_or_else(|| self.new_type_var().into());
+            self.declared_types.insert(name.clone(), ty);
+        }
+
+        // Now go through all symbols and derive types for the expressions.
+        // While analyzing a symbol, we ignore its declared type (unless the
+        // symbol is referenced). Unifying the declared type with the inferred
+        // type is done at the end.
+        for name in names {
+            // Ignore builtins (removed from definitions) and definitions without value.
+            let Some((_, Some(value))) = definitions.get_mut(&name) else {
+                continue;
+            };
+
+            let declared_type = self.declared_types[&name].clone();
+            (if declared_type.vars.is_empty() {
```

Comment:
where's this `(` closing? Can you rearrange this code a bit? It's kinda hard to read atm